File: examples/posix-thread/example01.cpp
// =================================================================
//
// File: example01.cpp
// Author: Pedro Perez
// Description: This file contains the code that adds all the
//				elements of an integer array using pthreads.
//              To compile: g++ example01.cpp -lpthread
//
// Copyright (c) 2022 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <iostream>
#include <iomanip>
#include <unistd.h>
#include <pthread.h>
#include "utils.h"

using namespace std;

const int SIZE = 100000000; //1e8
const int THREADS = 4;

typedef struct {
  int start, end; // [start, end)
  int *arr;
  double result;
} Block;

void* sumArray(void* param) {
  double acum;
  Block *block;
  int i;

  block = (Block *) param;
  acum =  0;
  for (i = block->start; i < block->end; i++) {
    acum += block->arr[i];
  }
  block->result = acum;
  pthread_exit(0);
}

int main(int argc, char* argv[]) {
	int *a, block_size, i, j;
	double ms, result;
	Block blocks[THREADS];
	pthread_t tids[THREADS];

	a = new int[SIZE];
	fill_array(a, SIZE);
	display_array("a", a);

  block_size = SIZE / THREADS;
  for (i = 0; i < THREADS; i++) {
      blocks[i].arr = a;
      blocks[i].start = i * block_size;
      blocks[i].end = (i != (THREADS - 1))? (i + 1) * block_size : SIZE;
	  blocks[i].result = 0;
  }

	std::cout << "Starting...\n";
	ms = 0;
    for (j = 0; j < N; j++) {
        start_timer();

        for (i = 0; i < THREADS; i++) {
            pthread_create(&tids[i], NULL, sumArray, (void*) &blocks[i]);
        }

		result = 0;
        for (i = 0; i < THREADS; i++) {
            pthread_join(tids[i], NULL);
            result += blocks[i].result;
        }

        ms += stop_timer();
    }
	std::cout << "sum = " << fixed << setprecision(0) << result << "\n";
	std::cout << "avg time =  " << setprecision(5) << (ms / N) << "\n";

	delete [] a;
	return 0;
}

File: examples/thread-python/example01.py
## =================================================================
##
## File: example01.py
## Author: Pedro Perez
## Description: This file contains the code that adds all the
##				elements of an integer array using threads in Python.
##
## Copyright (c) 2022 by Tecnologico de Monterrey.
## All Rights Reserved. May be reproduced for any non-commercial
## purpose.
##
## =================================================================

import utils
import time
import multiprocessing as mp
import threading as th

# CORES = mp.cpu_count() 
CORES = 4
SIZE = 100000000 ##1e8

def partialSumArray(id, start, end, array, results):
	acum = 0
	for i in range(start, end):
		acum += array[i]
	results[id] = acum

if __name__ == "__main__":
	array = [0] * SIZE
	results = [0] * CORES
	
	utils.fillArray(array)
	utils.displayArray("array", array)

	blockSize = SIZE // CORES
	
	print("Starting...")
	startTime = endTime = ms = 0
	for i in range(utils.N):
		startTime = time.time() * 1000
		
		threads = list()
		for i in range(CORES):
			start = i * blockSize
			if i != (CORES - 1):
				end = (i + 1) * blockSize
			else:
				end = SIZE
			t = th.Thread(target=partialSumArray, args=(i, start, end, array, results,))            
			threads.append(t)
			t.start()
		endTime = time.time() * 1000
		ms = ms + (endTime - startTime)

		result = 0
		for i in range(CORES):
			threads[i].join()
			result += results[i]

	print("sum = ", result)
	print("avg time = ", (ms / utils.N), " ms")